### PR TITLE
KER-351: Fix save hearing as copy

### PR DESCRIPTION
**Pull request by danipran**
TL;DR:
- Rename `cleanHearing` -> `prepareHearingForSave`
- Remove IDs from *all* objects when preparing hearing for saving, not just the section
- Prepare hearing only when creating a new hearing (either a fresh hearing or a copy of an existing hearing)
- Add new `reference_id` field in files/images on save
- Mark freshly uploaded files with `isNew`
- Don't assign uploaded files for hearing before saving

---

The "save as copy" action relies on some rather strange back-end logic where removing the ID is required only on some instances and, on the other hand, requires an ID in files. However, if you were creating a hearing the first time with a file uploaded to it, it causes problems.

To somewhat solve this problem, [a new field called `reference_id` can now be used with files and images](https://github.com/City-of-Helsinki/kerrokantasi/pull/490), which lets you base a file/image on an existing file/image (only when creating a new hearing, it currently does nothing if used for updating). This PR adds support for that.

Newly uploaded files are now marked with `isNew` to differentiate between existing instances and freshly uploaded files. This need for this comes up when you edit an existing hearing with a file and add a new file; now the hearing draft has both an existing file and a new file. If you save a copy in this state, the existing file needs to be copied while the new file doesn't need copying, you just want to assign it to the copy.

This PR also fixes another bug related to editing and uploading files; if you edit an existing hearing, add a file to a section and then leave without saving, the file is still added to the hearing despite the fact that the edits weren't saved. After the PR's changes, the hearing must be saved to finish adding the file to the hearing.

Related back-end PR (doesn't work without it): https://github.com/City-of-Helsinki/kerrokantasi/pull/490

### CODE REVIEW

**Comment by terovirtanen** [comment]
**KERROKANTASI-UI branch is deployed to platta: https://kerrokantasi-pr1025-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by danipran** [comment]
Ah yes, of course I realize my mistake *after* creating the pull request (no way to create an image or a file if there's no data nor reference to it). Back to the drawing board.

**Comment by terovirtanen** [comment]
**KERROKANTASI-UI branch is deployed to platta: https://kerrokantasi-pr1025-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## e2e tests result is __success__ for https://kerrokantasi-pr1025-ui.dev.hel.ninja :laughing::tada::tada::tada:

**Comment by terovirtanen** [comment]
**KERROKANTASI-UI branch is deployed to platta: https://kerrokantasi-pr1025-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## e2e tests result is __success__ for https://kerrokantasi-pr1025-ui.dev.hel.ninja :laughing::tada::tada::tada:

File: src/utils/hearingEditor.js [hunk 1]
```diff
@@ -230,16 +227,43 @@ export const parseCollection = (featureCollection) => {
...
   }, []);
   return { type: featureCollection.type, features: normalizedFeatures };
 };
 
-export const cleanHearing = (hearing) => {
-  let cleanedHearing = {};
+export const prepareSection = (section) => ({
+  ...section,
+  id: '',
+  images: section.images.reduce((images, image) => [...images, { ...image, id: '', reference_id: image.id }], []),
```
Comment (mikkojamG) [inline src/utils/hearingEditor.js:241]: I would actually go for map instead of reduce, just a tiny bit cleaner 🙂 
`section.images.map((image) => ({ ...image, id: '', reference_id: image.id }))`

File: src/utils/hearingEditor.js [hunk 2]
```diff
@@ -230,16 +227,43 @@ export const parseCollection = (featureCollection) => {
...
+  files: section.files.reduce(
```
Comment (mikkojamG) [inline src/utils/hearingEditor.js:242]:
```
section.files.map(
  (file) => ({ ...file, id: file.isNew ? file.id : '', reference_id: file.isNew ? '' : file.id })
)
```

File: src/utils/hearingEditor.js [hunk 3]
```diff
@@ -230,16 +227,43 @@ export const parseCollection = (featureCollection) => {
...
+    (files, file) => [...files, { ...file, id: file.isNew ? file.id : '', reference_id: file.isNew ? '' : file.id }],
+    [],
+  ),
+  questions: section.questions.reduce(
```
Comment (mikkojamG) [inline src/utils/hearingEditor.js:246]:
```
section.questions.map(
  (question) =>
  ({
    ...question,
    id: '',
    options: question?.options.map((option) => ({ ...option, id: '' })),
  })
)
```

File: src/utils/hearingEditor.js (hunk 1 above)
Comment (danipran) [reply src/utils/hearingEditor.js:241]: Agreed 👍

**Comment by terovirtanen** [comment]
**KERROKANTASI-UI branch is deployed to platta: https://kerrokantasi-pr1025-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**KERROKANTASI-UI branch is deployed to platta: https://kerrokantasi-pr1025-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## e2e tests result is __failed__ for https://kerrokantasi-pr1025-ui.dev.hel.ninja :crying_cat_face::anger::boom::boom:

**Comment by terovirtanen** [comment]
**KERROKANTASI-UI branch is deployed to platta: https://kerrokantasi-pr1025-ui.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## e2e tests result is __success__ for https://kerrokantasi-pr1025-ui.dev.hel.ninja :laughing::tada::tada::tada:

**Review by mikkojamG — APPROVED** [review]
